tests/opera/test_csar.py

```diff
@@ -0,0 +1,78 @@
+from opera.error import CsarValidationError, UnsupportedToscaFeatureError, FileOutOfBoundsError
+
+_RESOURCE_DIRECTORY = pathlib.Path(__file__).parent.parent.absolute() / "resources/"
+
+
+class CsarTestMode(enum.Enum):
+    ZIP = 1
+    DIRECTORY = 2
+    VIRTUAL = 3
+
+
+# https://stackoverflow.com/questions/1855095/
+def _zip_directory_contents(contents_path: pathlib.Path, destination_file_path: str):
+    with zipfile.ZipFile(destination_file_path, mode="a") as zipf:
+        for root, _, files in os.walk(str(contents_path)):
+            # don't include the root directory as the dot
+            relative_dir_path = os.path.relpath(root, str(contents_path))
+            if relative_dir_path != ".":
+                zipf.write(root, arcname=relative_dir_path)
+            for file in files:
+                zipf.write(os.path.join(root, file),
+                           arcname=os.path.relpath(os.path.join(root, file), str(contents_path)))
+
+
+def _base_loader(path: pathlib.Path, mode: CsarTestMode, tmp_path: pathlib.Path):
+    if mode == CsarTestMode.ZIP:
+        _, tmpfile_path = tempfile.mkstemp(prefix="xoperatmp-", suffix=".zip", dir=str(tmp_path))
+        _zip_directory_contents(path, tmpfile_path)
+        ToscaCsar.load(tmpfile_path)
+
+        # on success, the temp file is deleted
+        os.remove(tmpfile_path)
+    elif mode == CsarTestMode.DIRECTORY:
+        ToscaCsar.load(str(path))
+    else:
+        raise Exception("Unsupported test CSAR mode.")
+
+
+class TestCsar:
+    @pytest.mark.parametrize("mode", [CsarTestMode.ZIP, CsarTestMode.DIRECTORY])
+    def test_load_example_mini(self, mode, tmp_path: pathlib.Path):
+        _base_loader(_RESOURCE_DIRECTORY / "csar/mini/", mode, tmp_path)
+
+    @pytest.mark.parametrize("mode", [CsarTestMode.ZIP, CsarTestMode.DIRECTORY])
+    def test_load_barebones(self, mode, tmp_path: pathlib.Path):
+        _base_loader(_RESOURCE_DIRECTORY / "csar/barebones/", mode, tmp_path)
+
+    @pytest.mark.parametrize("mode", [CsarTestMode.ZIP, CsarTestMode.DIRECTORY])
+    def test_load_metadata_directory(self, mode, tmp_path: pathlib.Path):
+        with pytest.raises(UnsupportedToscaFeatureError):
+            _base_loader(_RESOURCE_DIRECTORY / "csar/metadatadir/", mode, tmp_path)
+
+    @pytest.mark.parametrize("mode", [CsarTestMode.ZIP, CsarTestMode.DIRECTORY])
+    def test_load_missing_metadata(self, mode, tmp_path: pathlib.Path):
+        with pytest.raises(CsarValidationError):
+            _base_loader(_RESOURCE_DIRECTORY / "csar/missingmetadata/", mode, tmp_path)
+
+    @pytest.mark.parametrize("mode", [CsarTestMode.ZIP, CsarTestMode.DIRECTORY])
+    def test_load_multiple_root_files(self, mode, tmp_path: pathlib.Path):
+        with pytest.raises(CsarValidationError):
+            _base_loader(_RESOURCE_DIRECTORY / "csar/multipleroot/", mode, tmp_path)
+
+    @pytest.mark.parametrize("where", ["../../../../../../../../../../../../../../../../etc/hosts", "/etc/hosts"])
```

Comment:
Could you also include `"../../etc/hosts"` which should also already fall out of the csar root path?

Reply:
I've added tests for non-existent files outside the CSAR, relatively closer in the tree, via relative and absolute paths, which I think you were aiming for.

Reply:
Actually, what I was aiming at is the files that _do_ exist outside the tree root or inside the peer trees.

Reply:
What do you mean? `../../etc/hosts` would not exist in the scope of this test, as it would point to `<resource-directory>/csar/barebones/../../etc/hosts -> <resource-directory>/etc/hosts`. I've used a (hopefully always valid) relative path and an absolute path to a file that, again hopefully, always exists. If the two hopefullies are what is bothering you, I've now explicitly created a temporary file that must exist and written separate tests for absolute and relative access.